### ISSUE TITLE
YCSB workload A kills cockroach

### ISSUE THREAD

**Issue by amq** [cockroachdb/cockroach#20448]
**BUG REPORT**

I can consistently replicate Cockroach getting into a deadlock situation by running YCSB [`workload A`](https://github.com/brianfrankcooper/YCSB/wiki/Core-Workloads)

> Workload A: Update heavy workload
> This workload has a mix of 50/50 reads and writes. An application example is a session store recording recent actions.

This is how I run it:

Install
```
wget https://github.com/brianfrankcooper/YCSB/releases/download/0.12.0/ycsb-0.12.0.tar.gz
tar -zxvf ycsb-0.12.0.tar.gz
cd ycsb-0.12.0
wget https://jdbc.postgresql.org/download/postgresql-42.1.4.jar
mv postgresql-42.1.4.jar lib
ant
```

Create schema
```
DROP DATABASE IF EXISTS ycsb;
CREATE DATABASE ycsb;
CREATE TABLE ycsb.usertable (
YCSB_KEY VARCHAR (255) PRIMARY KEY,
  FIELD0 TEXT,
  FIELD1 TEXT,
  FIELD2 TEXT,
  FIELD3 TEXT,
  FIELD4 TEXT,
  FIELD5 TEXT,
  FIELD6 TEXT,
  FIELD7 TEXT,
  FIELD8 TEXT,
  FIELD9 TEXT
);
```

Load data
```
./bin/ycsb load jdbc -P workloads/workloada -p db.driver=org.postgresql.Driver -p db.url=jdbc:postgresql://cockroach-01:26257/ycsb -p db.user=root -p db.batchsize=1000 -p recordcount=50000
```

Run
```
for threads in 1 2 4 8 16 32 64 96 128 160 192; do
  ./bin/ycsb run jdbc -P workloads/workloada -p db.driver=org.postgresql.Driver -p db.url=jdbc:postgresql://cockroach-01:26257/ycsb -p db.user=root -p operationcount=0 -p maxexecutiontime=60 -s -threads $threads
done
```

On m5.xlarge with 500 GB gp2 storage (1500 IOPS with bursts up to 3000) cockroach gets into a deadlock at around after `threads=96`. Please note that running a single test with `threads=96` is not sufficient, because the load accumulates with each run.

If deadlocks under a heavy load of updates is not considered a bug, please close the issue.

**Comment by petermattis** [comment]
@amq Deadlock refers to a locking pattern that prevents forward progress. Did you actually see deadlock (and if so, what locks are involved), or is it that throughput falls to 0? Zero throughput might be deadlock, or it might be something else. Regardless, it is very concerning.

What version of Cockroach are you running? How many Cockroach machines are you running?

**Comment by arjunravinarayan** [comment]
@amq, thank you for the report. i am attempting to reproduce this, thank you for the detailed reproduction instructions. I'll report back with the results of my reproduction.

**Comment by amq** [comment]
I had v1.1.3 in three-node cluster.

- throughput dropped to 0
- `show queries` showed dozens of queries which were more than an hour old, still executing

**Comment by tschottdorf** [comment]
@arjunravinarayan this could be similar to an issue a user reported privately. I'll send it your way (quota pool/snapshot refusal).

@amq any chance you're seeing message like these in the logs?

> W171130 22:48:34.342958 5631 storage/quota_pool.go:201  [n3,s2,r32/3:/Table/51{-/1/-3834…}] have been waiting 1m0.000147601s attempting to acquire quota

**Comment by amq** [comment]
I don't see `attempting to acquire quota`, but rather:
```
I171205 01:01:42.869333 149 server/status/runtime.go:223  [n1] runtime stats: 1.2 GiB RSS, 337 goroutines, 145 MiB/20 MiB/224 MiB GO alloc/idle/total, 738 MiB/881 MiB CGO alloc/total, 1037.69cgo/sec, 0.12/0.05 %(u/s)time, 0.00 %gc (0x)
W171205 01:01:46.040344 9234260 kv/dist_sender.go:1148  [client=10.0.0.172:56414,user=root,n1] have been waiting 1m0s sending RPC to r46 for batch: [txn: 070bf876], BeginTransaction [/Table/60/1/"user1573987489603120213"/0,/Min), Put [/Table/60/1/"user1573987489603120213"/0,/Min), EndTransaction [/Table/60/1/"user1573987489603120213"/0,/Min)
W171205 01:01:46.925540 9234258 kv/dist_sender.go:1148  [client=10.0.0.172:56460,user=root,n1] have been waiting 1m0s sending RPC to r46 for batch: [txn: 1b2f8854], BeginTransaction [/Table/60/1/"user1573987489603120213"/0,/Min), Put [/Table/60/1/"user1573987489603120213"/0,/Min), EndTransaction [/Table/60/1/"user1573987489603120213"/0,/Min)
W171205 01:01:47.409585 9234324 kv/dist_sender.go:1148  [client=10.0.0.172:56449,user=root,n1] have been waiting 1m0s sending RPC to r46 for batch: [txn: 513e0c60], BeginTransaction [/Table/60/1/"user1573987489603120213"/0,/Min), Put [/Table/60/1/"user1573987489603120213"/0,/Min), EndTransaction [/Table/60/1/"user1573987489603120213"/0,/Min)
W171205 01:01:50.761656 9234229 kv/dist_sender.go:1148  [client=10.0.0.172:56472,user=root,n1] have been waiting 1m0s sending RPC to r46 for batch: [txn: 68082f57], BeginTransaction [/Table/60/1/"user1573987489603120213"/0,/Min), Put [/Table/60/1/"user1573987489603120213"/0,/Min), EndTransaction [/Table/60/1/"user1573987489603120213"/0,/Min)
W171205 01:01:52.581574 9234216 kv/dist_sender.go:1148  [client=10.0.0.172:56392,user=root,n1] have been waiting 1m0s sending RPC to r46 for batch: [txn: a2dcccbf], BeginTransaction [/Table/60/1/"user1573987489603120213"/0,/Min), Put [/Table/60/1/"user1573987489603120213"/0,/Min), EndTransaction [/Table/60/1/"user1573987489603120213"/0,/Min)
<a minute later>
I171205 01:02:52.869334 149 server/status/runtime.go:223  [n1] runtime stats: 1.2 GiB RSS, 335 goroutines, 124 MiB/37 MiB/224 MiB GO alloc/idle/total, 762 MiB/889 MiB CGO alloc/total, 1021.70cgo/sec, 0.14/0.05 %(u/s)time, 0.00 %gc (1x)
W171205 01:02:53.443042 9234169 kv/dist_sender.go:1148  [client=10.0.0.172:56406,user=root,n1] have been waiting 1m0s sending RPC to r46 for batch: [txn: 6f207769], BeginTransaction [/Table/60/1/"user1573987489603120213"/0,/Min), Put [/Table/60/1/"user1573987489603120213"/0,/Min), EndTransaction [/Table/60/1/"user1573987489603120213"/0,/Min)
W171205 01:02:53.444220 9234215 kv/dist_sender.go:1148  [client=10.0.0.172:56492,user=root,n1] have been waiting 1m0s sending RPC to r46 for batch: [txn: c58413bb], BeginTransaction [/Table/60/1/"user1573987489603120213"/0,/Min), Put [/Table/60/1/"user1573987489603120213"/0,/Min), EndTransaction [/Table/60/1/"user1573987489603120213"/0,/Min)
W171205 01:02:53.446547 9234260 kv/dist_sender.go:1148  [client=10.0.0.172:56414,user=root,n1] have been waiting 1m0s sending RPC to r46 for batch: [txn: 08cc2705], BeginTransaction [/Table/60/1/"user1573987489603120213"/0,/Min), Put [/Table/60/1/"user1573987489603120213"/0,/Min), EndTransaction [/Table/60/1/"user1573987489603120213"/0,/Min)
W171205 01:02:59.962910 9234219 kv/dist_sender.go:1148  [client=10.0.0.172:56496,user=root,n1] have been waiting 1m0s sending RPC to r46 for batch: [txn: 25a8eac9], BeginTransaction [/Table/60/1/"user1573987489603120213"/0,/Min), Put [/Table/60/1/"user1573987489603120213"/0,/Min), EndTransaction [/Table/60/1/"user1573987489603120213"/0,/Min)
W171205 01:03:00.769275 9234341 kv/dist_sender.go:1148  [client=10.0.0.172:56580,user=root,n1] have been waiting 1m0s sending RPC to r46 for batch: [txn: 2e161520], BeginTransaction [/Table/60/1/"user1573987489603120213"/0,/Min), Put [/Table/60/1/"user1573987489603120213"/0,/Min), EndTransaction [/Table/60/1/"user1573987489603120213"/0,/Min)
W171205 01:03:00.769276 9234247 kv/dist_sender.go:1148  [client=10.0.0.172:56434,user=root,n1] have been waiting 1m0s sending RPC to r46 for batch: [txn: bb4b251d], BeginTransaction [/Table/60/1/"user1573987489603120213"/0,/Min), Put [/Table/60/1/"user1573987489603120213"/0,/Min), EndTransaction [/Table/60/1/"user1573987489603120213"/0,/Min)
W171205 01:03:00.774334 9234239 kv/dist_sender.go:1148  [client=10.0.0.172:56420,user=root,n1] have been waiting 1m0s sending RPC to r46 for batch: [txn: 3365dda9], BeginTransaction [/Table/60/1/"user1573987489603120213"/0,/Min), Put [/Table/60/1/"user1573987489603120213"/0,/Min), EndTransaction [/Table/60/1/"user1573987489603120213"/0,/Min)
```

In YCSB it looks like this:
```
2017-12-05 01:01:38:129 230 sec: 6024 operations; 0 current ops/sec; [READ: Count=0, Max=0, Min=9223372036854775807, Avg=?, 90=0, 99=0, 99.9=0, 99.99=0] [CLEANUP: Count=0, Max=0, Min=9223372036854775807, Avg=?, 90=0, 99=0, 99.9=0, 99.99=0] [UPDATE: Count=0, Max=0, Min=9223372036854775807, Avg=?, 90=0, 99=0, 99.9=0, 99.99=0]
Still waiting for thread Thread-98 to complete. Workload status: true
Still waiting for thread Thread-98 to complete. Workload status: true
Still waiting for thread Thread-98 to complete. Workload status: true
Still waiting for thread Thread-98 to complete. Workload status: true
Still waiting for thread Thread-98 to complete. Workload status: true
2017-12-05 01:01:48:129 240 sec: 6024 operations; 0 current ops/sec; [READ: Count=0, Max=0, Min=9223372036854775807, Avg=?, 90=0, 99=0, 99.9=0, 99.99=0] [CLEANUP: Count=0, Max=0, Min=9223372036854775807, Avg=?, 90=0, 99=0, 99.9=0, 99.99=0] [UPDATE: Count=0, Max=0, Min=9223372036854775807, Avg=?, 90=0, 99=0, 99.9=0, 99.99=0]
Still waiting for thread Thread-98 to complete. Workload status: true
Still waiting for thread Thread-98 to complete. Workload status: true
Still waiting for thread Thread-98 to complete. Workload status: true
Still waiting for thread Thread-98 to complete. Workload status: true
Still waiting for thread Thread-98 to complete. Workload status: true
```

**Comment by jordanlewis** [comment]
@solongordon the action item here is to see if it's possible to reproduce this issue with the app that's provided. You should be able to use a similar setup as the one you'll use for our internal ycsb load generator.

**Comment by solongordon** [comment]
I was able to reproduce this issue against v1.1.5 and it behaved just as @amq described. On v2.0, the behavior is somewhat better. I don't see the same unbounded accumulation of queries; there are only ever as many running queries as threads, and they're cleaned up immediately when the YCSB client exits.

However, I still see precipitous drops in qps and when running YCSB with a large number of threads (16 or more). I suspect this is related to it being an update-heavy workload with a zipfian distribution, which means multiple threads are likely to make concurrent updates to the same row. I'll dig in some more to see what's going on.

**Comment by solongordon** [comment]
The drop in qps definitely looks related to update contention, specifically for `user8425665148555282324`, the most common user in the distribution. I see some updates for that row taking upwards of a minute, which seems unreasonably high for the amount of load. I took a [session trace](https://gist.github.com/solongordon/aa3d7029997a614156b07372531d7d6c) of one such update.

Reading and writing from other rows is plenty fast while the workload is running.

**Comment by petermattis** [comment]
@nvanbenschoten or @tschottdorf Can you take a look at that trace? The query seems to be repeatedly running into conflicting intents on the key `/Table/53/1/"user8425665148555282324"/0`.

**Comment by nvanbenschoten** [comment]
One thing to point out is that the load generator is using the postgres jdbc driver without a [transaction retry loop](https://www.cockroachlabs.com/docs/stable/build-a-java-app-with-cockroachdb.html#transaction-with-retry-logic), so any retriable errors that are pushed to the client are presumably resulting in the transaction being thrown away (maybe not even rolled back/cleaned up). This would explain why each transaction ends up tripping over other transaction's intents and throwing `WriteIntentErrors`. It would also explain why no transaction is able to use intents it laid down at a previous epoch to make forward progress. Essentially, I think we're running into a livelock at higher concurrency levels because each transaction gets in the others' way for a while before throwing out all the work it performs after it hits a single retryable error.

@solongordon are you comfortable enough with this load generator to modify it? If so, could you try adding a transaction retry loop similar to the one [in our docs](https://www.cockroachlabs.com/docs/stable/build-a-java-app-with-cockroachdb.html#transaction-with-retry-logic)?

**Comment by bdarnell** [comment]
> One thing to point out is that the load generator is using the postgres jdbc driver without a transaction retry loop, so any retriable errors that are pushed to the client are presumably resulting in the transaction being thrown away (maybe not even rolled back/cleaned up).

We're *supposed* to clean up whenever a SQL transaction fails unless it contains a SAVEPOINT statement.

Also, the trace Solon posted showed one long request, so it's not just a cycle of unfair restarts caused by the lack of a within-transaction retry loop.

**Comment by nvanbenschoten** [comment]
> Also, the trace Solon posted showed one long request, so it's not just a cycle of unfair restarts caused by the lack of a within-transaction retry loop.

Yeah, it looks like you're right. The repeated `[Open pos:20]` logs indicate that we're in an open transaction that keeps getting retryable errors and [auto-retrying](https://github.com/cockroachdb/cockroach/blob/efb61119ca7a56d0aaab1552b81b0f147c8082bc/pkg/sql/conn_fsm.go#L264) without returning to the client.

The `PushTxn` requests are interesting in that trace. They each spend a long time (~200ms) waiting for conflicting commands in the command queue. Once out of the command queue, they seem to be successful in aborting the transaction they're trying to push. @solongordon it would be interesting to run with `--vmodule=cmd_push_txn=1` so that we can see what is logged [during the push txn evaluation](https://github.com/cockroachdb/cockroach/blob/efb61119ca7a56d0aaab1552b81b0f147c8082bc/pkg/storage/batcheval/cmd_push_txn.go#L203).

**Comment by solongordon** [comment]
I see a bunch of these in the leaseholder's log:
`I180314 15:52:13.530474 1409986 storage/batcheval/cmd_push_txn.go:216  [n3,s3,r33/3:/Table/58{-/1/"user…}] failed to push "sql txn" id=a569c6e3 key=/Table/58/1/"user1573987489603120213"/0 rw=true pri=0.09463374 iso=SERIALIZABLE stat=PENDING epo=1 ts=1521042732.342015627,1 orig=1521042718.978763840,1 max=1521042679.785790236,0 wto=false rop=false seq=15`

And these on the gateway node:
`W180314 15:53:57.859084 6398409 kv/dist_sender.go:1305  [client=10.142.0.58:49916,user=root,txn=68f574f7,n1] have been waiting 1m0s sending RPC to r33 (currently pending: [(n3,s3):3]) for batch: [txn: 68f574f7], BeginTransaction [/Table/58/1/"user1573987489603120213"/0,/Min), Put [/Table/58/1/"user1573987489603120213"/0,/Min)`

**Comment by nvanbenschoten** [comment]
@solongordon were there any logs from https://github.com/cockroachdb/cockroach/blob/efb61119ca7a56d0aaab1552b81b0f147c8082bc/pkg/storage/batcheval/cmd_push_txn.go#L208 You could try greping for `"pushee last active"`.

**Comment by solongordon** [comment]
Yes, should have mentioned, many of these as well (one per second or so):
`I180314 15:52:14.387612 1410237 storage/batcheval/cmd_push_txn.go:208  [n3,s3,r33/3:/Table/58{-/1/"user…}] 0983c622 pushed fafe4c62: pushee is expired (pushee last active: 1521042732.342015627,1)`

**Comment by petermattis** [comment]
@spencerkimball This is the issue I was mentioning. It touches retries and the push txn queue.

**Comment by nvanbenschoten** [comment]
Interesting, seeing a `pushee is expired` error without any nodes going down implies to me that a `TxnCoodSender` is abandoning a transaction record. 

@andreimatei you recently mentioned that you had found issues with how we roll back transactions after retryable errors. Could you expand on that? Do you think it could result in behavior like we see here, where transactions keep running into other abandoned transactions?

**Comment by andreimatei** [comment]
> @andreimatei you recently mentioned that you had found issues with how we roll back transactions after retryable errors. Could you expand on that? Do you think it could result in behavior like we see here, where transactions keep running into other abandoned transactions?

I don't think so. I was talking about cases where a txn's ctx is canceled. No reason to think that's happening here, I believe.

I've seen cases of expired transactions when I was overloading a cluster - the heartbeats weren't making it in time. I'd check whether that happens - heartbeat loops not running within 2s, or the heartbeat request taking a long time to execute.
When I've observed that in the past, it occurred to me that we should find a way to push back on... something when it occurs.

**Comment by tschottdorf** [comment]
Planning to take a look tomorrow, if there have any more insights please post them before that. ;-)

**Comment by bdarnell** [comment]
I'd also look at the fact that the txnWaitQueue unblocks all waiting requests at once, potentially creating thundering herds and making no attempts at fairness.

**Comment by nvanbenschoten** [comment]
I got the load generator up and running on a three-node roachprod cluster. I was quickly able to see throughput tank as concurrency increased. Thanks for making the reproduction steps so easy @amq!

First, I'll note that even when throughput is bottoming out, CPU utilization on all nodes is only at around 5% and memory use is well under under 1GB. This indicates that the problem is due to bad behavior in Cockroach and not due to an overloaded system.

It's also interesting to look at what the workload is doing when performance gets bad, because it isn't very much. YCSB is essentially just running the statement `UPDATE usertable SET field<N> = $1 WHERE ycsb_key = $2` over and over. `ycsb_key` is set using a zipf distribution, so a few rows will see a lot of traffic. The field number is chosen randomly between 0 and 9. When an UPDATE statement comes in to CRDB, it will find the KV with the matching `ycsb_key` and update the corresponding KV for the specified field.

Because the ddl above creates the table with a single column family, the UPDATE statement reduces to reading a key and then writing to that same key in a transaction. As concurrency increases in the transaction, performance drops. We certainly shouldn't be seeing such a bad drop in performance, but we do expect some hit as contention grows. If we are willing to change the schema, we can improve this situation significantly by breaking up the individual fields into multiple keys. We can do this by using [column families](https://www.cockroachlabs.com/docs/stable/column-families.html), resulting in the schema:

```sql
CREATE TABLE ycsb.usertable (
  YCSB_KEY VARCHAR (255) PRIMARY KEY,
  FIELD0 TEXT,
  FIELD1 TEXT,
  FIELD2 TEXT,
  FIELD3 TEXT,
  FIELD4 TEXT,
  FIELD5 TEXT,
  FIELD6 TEXT,
  FIELD7 TEXT,
  FIELD8 TEXT,
  FIELD9 TEXT,
  FAMILY f1 (YCSB_KEY),
  FAMILY f2 (FIELD0),
  FAMILY f3 (FIELD1),
  FAMILY f4 (FIELD2),
  FAMILY f5 (FIELD3),
  FAMILY f6 (FIELD4),
  FAMILY f7 (FIELD5),
  FAMILY f8 (FIELD6),
  FAMILY f9 (FIELD7),
  FAMILY f10 (FIELD8),
  FAMILY f11 (FIELD9)
);
```

This dramatically reduces the amount of write contention because even though all transactions still read from the same key, they now often write to different keys. Below is a comparison of throughput between a schema using a single column family and a schema using a column family per field. We can see that even with multiple column families, throughput is still impacted by concurrency. However, it does not drop to 0 like the setup with only one column family.

![chart](https://user-images.githubusercontent.com/5438456/37438303-ff516336-27c7-11e8-8df7-8177466a939c.png)

That improvement is nice, but such bad performance under contention shouldn't exist in the first place. I dug more through the logs and saw a lot of the following, which @solongordon pointed out above:

```
I180314 15:52:14.387612 1410237 storage/batcheval/cmd_push_txn.go:208 [n3,s3,r33/3:/Table/58{-/1/"user…}] 0983c622 pushed fafe4c62: pushee is expired (pushee last active: 1521042732.342015627,1)
```

This again showed that transaction coordinators were seemingly abandoning transaction records and failing to heartbeat them. I also saw the following:

```
I180315 00:27:07.392188 72328 kv/txn_coord_sender.go:1097  [client=10.142.0.29:39772,user=root,n1,txn=a58eb4e2] heartbeat failed: heartbeat for transaction "sql txn" id=a58eb4e2 key=/Table/52/1/"user1573987489603120213"/0 rw=false pri=0.03439626 iso=SERIALIZABLE stat=PENDING epo=1 ts=1521073616.475806071,1 orig=1521073616.475806071,1 max=1521073614.325326222,0 wto=false rop=false seq=13 failed; record not present
W180314 23:00:59.157659 224092 kv/txn_coord_sender.go:1026  [n1] abort due to inactivity failed for "sql txn" id=1c6c653a key=/Table/52/1/"user1573987489603120213"/0 rw=false pri=0.07864293 iso=SERIALIZABLE stat=PENDING epo=1 ts=1521068454.577385592,1 orig=1521068454.577385592,1 max=1521068451.585962884,0 wto=false rop=false seq=9: txn "sql txn" id=1c6c653a key=/Table/52/1/"user1573987489603120213"/0 rw=false pri=0.07864293 iso=SERIALIZABLE stat=PENDING epo=1 ts=1521068454.577385592,1 orig=1521068454.577385592,1 max=1521068451.086758817,0 wto=false rop=false seq=10: does not exist 
```

This is interesting because it shows that the txn_coord_sender is attempting to heartbeat its transaction record but failing because the record does not exist at the time that the heartbeat request is evaluated. I think this means that the heartbeat arrived at the txn record after a PushTxnRequest succeeded in aborting it. This isn't good. I'm wondering if its possible that the `HeartbeatTxnRequest` for the transactions that are successful in laying down intents are being starved by the attack of `PushTxnRequests` and `QueryTxnRequest` sent by contending transactions. All three of these request types would need to wait for each other in the `CommandQueue`, so it's possible that the `HeartbeatTxnRequest` is being delayed there so long that its record appears expired.

**Comment by tschottdorf** [comment]
The transaction heartbeat timeout is quite aggressive, so I wouldn't be surprised.

**Comment by bdarnell** [comment]
> The transaction heartbeat timeout is quite aggressive, so I wouldn't be surprised.

For the record, we send txn heartbeats on a 1-second interval, and consider a transaction expired if it hasn't heartbeat in 2 seconds (these values were reduced from 5/10s shortly before 1.0 shipped. cc @spencerkimball). That's aggressive, but not completely crazy. It would be reasonable to expect heartbeats to complete every second *if* we could prioritize them in the command queue so they didn't have to wait in line before all the attempts to push them.

**Comment by nvanbenschoten** [comment]
> For the record, we send txn heartbeats on a 1-second interval, and consider a transaction expired if it hasn't heartbeat in 2 seconds (these values were reduced from 5/10s shortly before 1.0 shipped.

I bumped up this txn expiration length back to 5 seconds but this didn't seem to help.

I am seeing something that really doesn't look right to me. Nowhere in the write cmd evaluation/proposal path do we ever filter out no-op commands or commands that decided that they didn't need to write anything during evaluation. Unless I'm missing something, this omission would be pretty catastrophic for this workload. The reason is that a contended workload will naturally result in waves of `PushTxnRequests` and `ResolveIntentRequests`. However, in most cases, no more than one of these requests should ever actually perform a write. The rest will usually find a transaction in whatever state it desires and should exit early. An example of this is a `PushTxnRequests` who finds its transaction in an ABORTED or COMMITTED state. This [should be a quick no-op](https://github.com/cockroachdb/cockroach/blob/05df77a3e403cba8e3c8e1d40c7d0de0d76cf191/pkg/storage/batcheval/cmd_push_txn.go#L157), but I think it will actually result in a consensus write. If each request in these waves actually performs a write, it's more understandable why `HeartbeatTxnRequests` are being starved in the `CommandQueue`.

All that said, I tried to fix this and did not see the results I was looking for, so I may either be wrong about this or the fundamental issue here may be unrelated.

**Comment by bdarnell** [comment]
> I am seeing something that really doesn't look right to me. Nowhere in the write cmd evaluation/proposal path do we ever filter out no-op commands or commands that decided that they didn't need to write anything during evaluation. Unless I'm missing something, this omission would be pretty catastrophic for this workload. 

Right. If the command returns an error we skip consensus and don't write anything, but we will still go through the consensus path for a successful no-op write (We don't have many of these, but PushTxn is one). It's not a complete no-op because we still update the raft applied index, but that's not strictly necessary (we don't do it for errors) and it's certainly not worth paying the consensus price here. 

> All that said, I tried to fix this and did not see the results I was looking for, so I may either be wrong about this or the fundamental issue here may be unrelated.

Where were you detecting the no-op? The raft applied index may have already been written to the batch.

**Comment by nvanbenschoten** [comment]
> Where were you detecting the no-op?

I was checking if `result.Result.Replicated == nil && len(result.Result.WriteBatch.Data) == 0`. It looked like the latter condition was never true, but I wasn't able to track down where the batch was being written to. It's possible that batches always have a header or something that I'll need to ignore. I tried to address that by iterating over the keys in the batch to check if it was empty, but I wasn't able to get that to work either. I'll continue trying today.

**Comment by nvanbenschoten** [comment]
After ~3~ ~4~ 5 false alarms, I think I figured out the primary issue here. The problem is related to the `txn_coord_sender` heartbeat loop. It turns out that the heartbeat loop indirectly [aborts a transaction](https://github.com/cockroachdb/cockroach/blob/a71acfd1ef257e7969551e6e75225644f457f9d0/pkg/kv/txn_coord_sender.go#L1103) and shuts down the heartbeat loop if it ever sees an error. This turns out to be an issue for transaction retries because of the criteria we use to start the heartbeat loop - that a transaction [has written intents](https://github.com/cockroachdb/cockroach/blob/a71acfd1ef257e7969551e6e75225644f457f9d0/pkg/kv/txn_coord_sender.go#L1301) sometime in the past.

The problem is that "the past" may mean a previous transaction epoch. It also may mean speculative intents, which may be [added to a `TxnCoordMeta`](https://github.com/cockroachdb/cockroach/blob/a71acfd1ef257e7969551e6e75225644f457f9d0/pkg/kv/txn_coord_sender.go#L466) in the presence of an `EndTransaction` batch even before the batch has succeeded. Combined, this breaks the guarantee that `len(meta.Intents) > 0` implies that a transaction record has ever even been written. The following is what we see over and over again in this scenerio:

```
1. txn issues a Scan request
2. txn issues a BeginTxn, Put, EndTxn
-- note that because of the EndTxn, we fall into the speculative intent case above
3. transaction is forced to restart due to a WriteTooOld error
4. refresh span request fails, restart propagates up to ConnExecutor
5. txn issues a second Scan request
6. response succeeds, `len(meta.Intents) > 0` from speculative intents, heartbeat loop starts
7. txn issues a second BeginTxn, Put, EndTxn
8. heartbeat loop fails to find a txn record, shuts down txn heartbeat loop
9. transaction record then gets written and intent is laid down but not committed. WriteTooOld error on EndTxn evaluation. Another retry
-- I'm not quite sure how this happens
10. retry never happens because it is blocked by maybeRejectClientLocked
11. other transactions block in txnwait queue for 2 seconds until the transaction record expires
```

I've gone back and forth on whether the issue here is the speculative intents being used to justify the heartbeat loop or whether it is the heartbeat loop aggressively shutting down when it doesn't find a transaction record. At the moment, I think the issue is the latter. The reason for this is that we start the heartbeat loop in [other cases](https://github.com/cockroachdb/cockroach/blob/a71acfd1ef257e7969551e6e75225644f457f9d0/pkg/kv/txn_coord_sender.go#L1265) where we're not positive that a transaction record exists as well. If aren't guaranteeing that a transaction record exists before starting the heartbeat loop then it shouldn't react the way it does when one is missing. The good news is that either of the following changes fixes this issue if combined with prototype fixes for #23942, #23943, and #23945 (seriously, it doesn't seem to work without all of them).

```
diff --git a/pkg/kv/txn_coord_sender.go b/pkg/kv/txn_coord_sender.go
index c37041cfce..a4864ba8f0 100644
--- a/pkg/kv/txn_coord_sender.go
+++ b/pkg/kv/txn_coord_sender.go
@@ -1290,8 +1290,8 @@ func (tc *TxnCoordSender) updateState(
        // set in these cases, it may be possible that the request was read-only.
        // This is ok, since the following block will be a no-op if the batch
        // contained no transactional write requests.
-       _, ambiguousErr := pErr.GetDetail().(*roachpb.AmbiguousResultError)
-       if pErr == nil || ambiguousErr || newTxn.Writing {
+       // _, ambiguousErr := pErr.GetDetail().(*roachpb.AmbiguousResultError)
+       if /*pErr == nil || ambiguousErr ||*/ newTxn.Writing {
                // Adding the intents even on error reduces the likelihood of dangling
                // intents blocking concurrent writers for extended periods of time.
                // See #3346.
```
or
```
diff --git a/pkg/kv/txn_coord_sender.go b/pkg/kv/txn_coord_sender.go
index c37041cfce..5212caf596 100644
--- a/pkg/kv/txn_coord_sender.go
+++ b/pkg/kv/txn_coord_sender.go
@@ -1100,6 +1101,11 @@ func (tc *TxnCoordSender) heartbeat(ctx context.Context) bool {
                        tc.mu.meta.Txn.Update(errTxn)
                        tc.mu.Unlock()
                }
+
+               if strings.Contains(pErr.Message, "record not present") {
+                       return true
+               }
+
                // We're not going to let the client carry out additional requests, so
                // try to clean up if the known txn disposition remains PENDING.
                if txn.Status == roachpb.PENDING {
```

![chart](https://user-images.githubusercontent.com/5438456/37505857-601000f6-28bd-11e8-82f0-a71d6ee8fe8c.png)

In addition to this, I found a few smaller problems while exploring this bug:
- We require consensus even for writes that are no-ops #23942
- ResolveIntent requests should not blindly write to the AbortSpan #23943
- QueryTxn requests declare the wrong key #23944
- BeginTxnRequest can be delayed long enough to expire and be immediately aborted #23945

**Comment by bdarnell** [comment]
Wow, nice detective work.

> 9. transaction record then gets written and intent is laid down but not committed. WriteTooOld error on EndTxn evaluation. Another retry
> -- I'm not quite sure how this happens
> 10. retry never happens because it is blocked by maybeRejectClientLocked

You're not quite sure *that* it's maybeRejectClientLocked, or you're not quite sure why maybeRejectClientLocked is triggering?

> (seriously, it doesn't seem to work without all of them).

Do you get no benefit at all without the full set of changes, or do you mean that they all provide incremental benefits?

> +       if /*pErr == nil || ambiguousErr ||*/ newTxn.Writing {

This breaks the cases described in the comment above, right?

I'd like to propose a third alternative for discussion: Add the full txn proto to HeartbeatTxnRequest and allow it to write a transaction record when none exists. This is similar to ignoring the "record not present" errors in the heartbeat loop, but it makes the heartbeats work (blocking other pushers) instead of silently ignoring failures. This feels consistent with the way we handle multi-range batches, by allowing other writes to happen before the BeginTransaction, although it does have some risk of writing a late transaction record that will just have to be GC'd. It would almost solve #23945, except that the HeartbeatTxn would be stuck in the command queue behind the BeginTransaction batch anyway. Maybe there's some way we could make the command queue smarter about txn ops. 

I think that for an immediate fix ignoring the errors is probably best. In the long run I'd look at the third option and refactor the relationship between BeginTxn and HeartbeatTxn.

**Comment by nvanbenschoten** [comment]
> You're not quite sure that it's maybeRejectClientLocked, or you're not quite sure why maybeRejectClientLocked is triggering?

What I wrote was unclear. I'm sure about both of those things. I'm not sure how the BeginTxn and Put requests are being evaluated and succeeding separately from the EndTxn, which is hitting a WriteTooOld error. They were previously being evaluated as a single batch, which is where my confusion is arising from.

> Do you get no benefit at all without the full set of changes, or do you mean that they all provide incremental benefits?

We see roughly the same "deadlock" when any one of them is missing. I think we're seeing this non-linear behavior where all of the changes are needed because with any one of them missing the first heartbeat to a new transaction record is being delayed long enough that the record expires.

> This breaks the cases described in the comment above, right?

This avoids the issue I was seeing, yes. I don't think it's actually the solution we want though, and I think you agree.

> I'd like to propose a third alternative for discussion: Add the full txn proto to HeartbeatTxnRequest and allow it to write a transaction record when none exists.

If we're going to move in this direction then I'd like to see if we could expand it all the way to the natural conclusion of deprecating `HeartbeatTxnRequest` and making `BeginTransactionRequest` subsume its responsibility entirely. `BeginTransactionRequest` would then need to be made idempotent, which would affect some of the replay protection it currently provides. This might be ok though. @tschottdorf may have thoughts on this, as I think he's been thinking about idempotency lately.

> This feels consistent with the way we handle multi-range batches, by allowing other writes to happen before the BeginTransaction

Could you expand on this? Are you referring to how we write an aborted transaction record when a PushTxnRequest is sent for a transaction that does not yet have a transaction record?

> It would almost solve #23945, except that the HeartbeatTxn would be stuck in the command queue behind the BeginTransaction batch anyway.

This would only be the case if we began the heartbeat when we sent the BeginTxn request instead of when we got the first response after sending it.

> Maybe there's some way we could make the command queue smarter about txn ops.

Yeah, I was beginning to think about the same thing. It would be really nice if we had some notion of priority in the CommandQueue so that HeartbeatTxn requests could jump in front of other lower priority requests.

**Comment by tschottdorf** [comment]
@nvanbenschoten the txn record is written the second time around because after the Epoch increment, it hits a [new path in DistSender](https://github.com/cockroachdb/cockroach/blob/146ad9e967c872ac1bd5560b6ea70d5dc4ba3c5a/pkg/kv/dist_sender.go#L633) which splits off the `EndTransaction` into a separate batch (to prevent starvation, ironically).

The `Writing` status is generally a mess, but in the 2.0.x time frame, wouldn't making the heartbeat loop accept the absence of the txn record (at least for the first heartbeat) solve the problem? I don't think we ever rely on the heartbeat loop aborting in that scenario.

In the longer term, I agree that it'd be nice if `BeginTransaction` and `HeartbeatTxn` fused into one, but also we'd want to get rid of the `Writing` flag, and there are lots of subtle considerations to be made (I'm fleshing up a few of those in a related upcoming tech note).

**Comment by bdarnell** [comment]
> > This breaks the cases described in the comment above, right?

> This avoids the issue I was seeing, yes. I don't think it's actually the solution we want though, and I think you agree.

"The comment above" referred to the [comment in the code above this line](https://github.com/cockroachdb/cockroach/blob/9fe8102332fc67d1a364332da69f1c0ea582b929/pkg/kv/txn_coord_sender.go#L1274-L1282). This change would break things for that scenario. 

> If we're going to move in this direction then I'd like to see if we could expand it all the way to the natural conclusion of deprecating HeartbeatTxnRequest and making BeginTransactionRequest subsume its responsibility entirely. BeginTransactionRequest would then need to be made idempotent, which would affect some of the replay protection it currently provides. This might be ok though. @tschottdorf may have thoughts on this, as I think he's been thinking about idempotency lately.

Yeah, allowing HeartbeatTxn to write a txn record would be equivalent to making BeginTransaction replayable. One reason why we might want to keep both methods is to switch from BeginTxn to HeartbeatTxn after we know that the txn record has been written once, since we'd know that any subsequent missing txn record would be because it had been aborted and cleaned up. (HeartbeatTxn might also be a little cheaper since it wouldn't need to carry the TxnMeta data)

> This would only be the case if we began the heartbeat when we sent the BeginTxn request instead of when we got the first response after sending it.

How so? The batch (BeginTxn, Put) could be blocked in the command queue on the put's key, but that batch will then block *all* of its keys. The heartbeat couldn't go through until the Begin/Put batch finishes, even if we've gotten responses from other parts of the split batch.
